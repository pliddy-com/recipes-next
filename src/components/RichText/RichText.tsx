import { ReactNode } from 'react';

import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';

import {
  documentToReactComponents,
  Options,
} from '@contentful/rich-text-react-renderer';

import { BLOCKS, INLINES, MARKS } from '@contentful/rich-text-types';

import {
  RecipeDescription,
  RecipeDescriptionLinks,
} from 'types/generated/graphql';

type BlockProps = {
  children: ReactNode | ReactNode[];
};

const P = ({ children }: BlockProps) => (
  <Typography variant="body1" paragraph>
    {children}
  </Typography>
);

const H1 = ({ children }: BlockProps) => (
  <Typography variant="h1">{children}</Typography>
);

const H2 = ({ children }: BlockProps) => (
  <Typography variant="h2">{children}</Typography>
);

const H3 = ({ children }: BlockProps) => (
  <Typography variant="h3">{children}</Typography>
);

const H4 = ({ children }: BlockProps) => (
  <Typography variant="h4">{children}</Typography>
);

const H5 = ({ children }: BlockProps) => (
  <Typography variant="h5">{children}</Typography>
);

const H6 = ({ children }: BlockProps) => (
  <Typography variant="h6">{children}</Typography>
);

type RenderOptionProps = {
  links?: RecipeDescriptionLinks;
};

const renderOptions = ({ links }: RenderOptionProps): Options => {
  const { entries } = links ?? {};
  const { inline } = entries ?? {};

  // create an entry map
  const entryMap = new Map();

  if (inline) {
    for (const entry of inline) {
      entryMap.set(entry?.sys?.id, entry);
    }
  }

  return {
    renderMark: {
      [MARKS.BOLD]: (text) => <strong>{text}</strong>,
    },
    renderNode: {
      [BLOCKS.PARAGRAPH]: (_node, children) => <P>{children}</P>,
      [BLOCKS.HEADING_1]: (_node, children) => <H1>{children}</H1>,
      [BLOCKS.HEADING_2]: (_node, children) => <H2>{children}</H2>,
      [BLOCKS.HEADING_3]: (_node, children) => <H3>{children}</H3>,
      [BLOCKS.HEADING_4]: (_node, children) => <H4>{children}</H4>,
      [BLOCKS.HEADING_5]: (_node, children) => <H5>{children}</H5>,
      [BLOCKS.HEADING_6]: (_node, children) => <H6>{children}</H6>,
      [INLINES.EMBEDDED_ENTRY]: (node) => {
        const { title, slug } = entryMap.get(node.data.target.sys.id);
        return <a href={slug}>{`${title} `}</a>;
      },
    },
  };
};

// NOTE: is being passed json object; needs to know type/structure of full rich text payload
//       and parse json and links children

type RichTextProps = {
  content: RecipeDescription;
};

export const RichText = ({ content }: RichTextProps) => {
  const { json, links } = content ?? {};

  return <Box>{documentToReactComponents(json, renderOptions({ links }))}</Box>;
};

export default RichText;
