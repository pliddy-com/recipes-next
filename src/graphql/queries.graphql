# Query for Homepage RecipeCollection (all recipes)

query homePageQuery {
  recipeCollection(order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe collection based on filter

query recipeCollection($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe page

query recipePage($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC, limit: 1) {
    total
    items {
      ...RecipeDefault
    }
  }
}

# Query for List Page RecipeCollection (all recipes matching tag)

query listPageQuery($where: TagFilter) {
  tagCollection(where: $where, limit: 1) {
    total
    items {
      slug
      title
      linkedFrom {
        recipeCollection(limit: 24) {
          total
          items {
            ...RecipeSummary
          }
        }
      }
    }
  }
}

# Query for recipe slugs

query recipeSlugsCollection($where: RecipeFilter) {
  recipeCollection(where: $where) {
    items {
      slug
    }
  }
}

# query for category slugs

query categorySlugsCollection($where: TagFilter) {
  tagCollection(where: $where) {
    items {
      slug
    }
  }
}

# query for category slugs

query tagSlugsCollection($where: TagFilter) {
  tagCollection(where: $where) {
    items {
      slug
      linkedFrom {
        recipeCollection {
          total
        }
      }
    }
  }
}

# query for Taxonomy with Category & Subcategory levels

query taxonomyCollection($where: TaxonomyFilter) {
  taxonomyCollection(where: $where, limit: 1) {
    items {
      ...TaxonomyDefault

      tag {
        ...TagDefault
      }

      childrenCollection(limit: 24) {
        total
        items {
          ... on Tag {
            ...TagDefault

            linkedFrom {
              recipeCollection(limit: 1) {
                total
              }
            }
          }
          ...TaxonomyWithChildren
        }
      }
    }
  }
}
