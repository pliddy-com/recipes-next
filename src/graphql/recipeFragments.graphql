# Fragments with default fields for Recipe-specific sections:
#   * Ingredients
#   * Instructions
#   * Image
#   * Recipe

fragment IngredientsDefault on IngredientSection {
  sys {
    id
  }
  title
  slug
  label
  ingredientList
}

fragment InstructionsDefault on InstructionSection {
  sys {
    id
  }
  title
  slug
  label
  instructionList
}

fragment ImageDefault on Asset {
  sys {
    id
  }
  __typename
  title
  description
  contentType
  fileName
  size
  url
  height
  width
}

fragment DescriptionDefault on Recipe {
  description {
    json
    links {
      entries {
        inline {
          __typename
          sys {
            id
            __typename
          }
        }
        hyperlink {
          __typename
          sys {
            id
            __typename
          }
        }
        block {
          __typename

          sys {
            id
            __typename
          }
        }
      }
      assets {
        hyperlink {
          __typename
          title
        }
        block {
          __typename
          title
        }
      }
    }
  }
}

fragment RecipeDefault on Recipe {
  sys {
    id
  }
  __typename
  title
  slug
  description {
    json
  }
  abstract
  image {
    ...ImageDefault
  }
  ingredientsCollection(limit: 10) {
    items {
      ...IngredientsDefault
    }
  }
  equipment
  instructionsCollection(limit: 10) {
    items {
      ...InstructionsDefault
    }
  }
  notes
  tagsCollection(limit: 10) {
    items {
      ...TagDefault
    }
  }
}
