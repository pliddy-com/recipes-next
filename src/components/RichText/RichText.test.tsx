// import testing-library methods
import { render } from '@testing-library/react';

// add custom jest matchers from jest-dom
import '@testing-library/jest-dom';

// import the component to test
import RichText from 'components/RichText/RichText';

import { RecipeDescription } from 'types/generated/graphql';

describe('RichText', () => {
  const content = {
    json: {
      nodeType: 'document',
      data: {},
      content: [
        {
          nodeType: 'paragraph',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: 'This is a normal paragraph of text.',
              marks: [],
              data: {},
            },
          ],
        },
        {
          nodeType: 'paragraph',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: '',
              marks: [],
              data: {},
            },
            {
              nodeType: 'hyperlink',
              data: {
                uri: 'https://recipes.pliddy.com',
              },
              content: [
                {
                  nodeType: 'text',
                  value: 'This is an external link.',
                  marks: [],
                  data: {},
                },
              ],
            },
            {
              nodeType: 'text',
              value: '',
              marks: [],
              data: {},
            },
          ],
        },
        {
          nodeType: 'heading-1',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: 'This is a level 1 heading.',
              marks: [],
              data: {},
            },
          ],
        },
        {
          nodeType: 'heading-2',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: 'This is a level 2 heading.',
              marks: [],
              data: {},
            },
          ],
        },
        {
          nodeType: 'heading-3',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: 'This is a level 3 heading.',
              marks: [],
              data: {},
            },
          ],
        },
        {
          nodeType: 'heading-4',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: 'This is a level 4 heading.',
              marks: [],
              data: {},
            },
          ],
        },
        {
          nodeType: 'heading-5',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: 'This is a level 5 heading.',
              marks: [],
              data: {},
            },
          ],
        },
        {
          nodeType: 'heading-6',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: 'This is a level 6 heading.',
              marks: [],
              data: {},
            },
          ],
        },
        {
          nodeType: 'paragraph',
          data: {},
          content: [
            {
              nodeType: 'text',
              value: '',
              marks: [],
              data: {},
            },
            {
              nodeType: 'embedded-entry-inline',
              data: {
                target: {
                  sys: {
                    id: '1Z5PsRKqUsxYrWy4tI9jqs',
                    type: 'Link',
                    linkType: 'Entry',
                  },
                },
              },
              content: [],
            },
            {
              nodeType: 'text',
              value: '',
              marks: [],
              data: {},
            },
          ],
        },
      ],
    },
    links: {
      assets: {
        block: [],
        hyperlink: [],
        __typename: 'RecipeDescriptionAssets',
      },
      entries: {
        inline: [
          {
            __typename: 'Recipe',
            sys: {
              id: '1Z5PsRKqUsxYrWy4tI9jqs',
              environmentId: 'abab825b-f415-453b-afff-f68b7fa1d712',
              spaceId: 'fo9qwg6zarbt',
              __typename: 'Sys',
            },
            slug: 'red-beans-and-rice',
            title: 'Red Beans & Rice',
          },
        ],
        hyperlink: [],
        __typename: 'RecipeDescriptionEntries',
      },
      __typename: 'RecipeDescriptionLinks',
    },
    __typename: 'RecipeDescription',
  };

  it('renders the rich text section if there is content', async () => {
    const expectedText = content.json.content[0].content[0].value;

    const expectedExternalText =
      content.json.content[1].content[1].content?.[0].value;

    const { uri: expectedExternalHref } = content.json.content[1].content[1]
      .data as {
      uri: string;
    };
    const expectedRefText = content.links.entries.inline[0].title;
    const expectedRefHref = content.links.entries.inline[0].slug;

    const { getByRole, getByText } = render(
      <RichText content={content as unknown as RecipeDescription} />
    );

    const p = getByText(expectedText as string);
    const h1 = getByRole('heading', { level: 1 });
    const h2 = getByRole('heading', { level: 2 });
    const h3 = getByRole('heading', { level: 3 });
    const h4 = getByRole('heading', { level: 4 });
    const h5 = getByRole('heading', { level: 5 });
    const h6 = getByRole('heading', { level: 6 });
    const externalLink = getByRole('link', {
      name: expectedExternalText,
    });
    const refLink = getByRole('link', {
      name: expectedRefText,
    });

    expect(p).toBeInTheDocument();
    expect(h1).toBeInTheDocument();
    expect(h2).toBeInTheDocument();
    expect(h3).toBeInTheDocument();
    expect(h4).toBeInTheDocument();
    expect(h5).toBeInTheDocument();
    expect(h6).toBeInTheDocument();
    expect(externalLink).toBeInTheDocument();
    expect(externalLink).toHaveAttribute('href', expectedExternalHref);
    expect(refLink).toBeInTheDocument();
    expect(refLink).toHaveAttribute('href', expectedRefHref);
  });

  it('does not render the rich text section if there is no content', () => {
    const content = undefined;
    const { queryByTestId } = render(
      <RichText content={content as unknown as RecipeDescription} />
    );

    const richText = queryByTestId('richtext');
    expect(richText).toBeNull();
  });

  it('handles rendering if there are no entries in links', () => {
    const content = {
      json: {
        nodeType: 'document',
        data: {},
        content: [],
      },
      links: {},
    };

    const { queryByTestId } = render(
      <RichText content={content as unknown as RecipeDescription} />
    );

    const richText = queryByTestId('richtext');
    expect(richText).toBeInTheDocument();
  });

  it('handles rendering if there are no links', () => {
    const content = {
      json: {
        nodeType: 'document',
        data: {},
        content: [],
      },
    };

    const { queryByTestId } = render(
      <RichText content={content as unknown as RecipeDescription} />
    );

    const richText = queryByTestId('richtext');
    expect(richText).toBeInTheDocument();
  });
});
