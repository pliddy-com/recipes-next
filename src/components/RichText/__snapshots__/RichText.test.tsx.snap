// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichText when there is content renders the component 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
    data-testid="richtext"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-z2eky3-MuiTypography-root"
    >
      This is a normal paragraph of text.
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-z2eky3-MuiTypography-root"
    >
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
        href="https://recipes.pliddy.com"
        rel="noreferrer"
        target="_blank"
      >
        This is an external link.
      </a>
    </p>
    <h1
      class="MuiTypography-root MuiTypography-h1 css-o2w69a-MuiTypography-root"
    >
      This is a level 1 heading.
    </h1>
    <h2
      class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
    >
      This is a level 2 heading.
    </h2>
    <h3
      class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
    >
      This is a level 3 heading.
    </h3>
    <h4
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      This is a level 4 heading.
    </h4>
    <h5
      class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
    >
      This is a level 5 heading.
    </h5>
    <h6
      class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
    >
      This is a level 6 heading.
    </h6>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-z2eky3-MuiTypography-root"
    >
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
        href="red-beans-and-rice"
      >
        Red Beans & Rice
      </a>
    </p>
  </div>
</div>
`;
